sql/opt_hints.cc

```diff
@@ -398,6 +401,35 @@ Opt_hints_qb::Opt_hints_qb(Opt_hints *opt_hints_arg,
 }
 
 
+void Opt_hints_qb::fix_hints_for_table(TABLE_LIST *table_list)
+{
+  Opt_hints_table *tab= static_cast<Opt_hints_table *>(find_by_name(table_list->alias));
+
+  /*
+    If this is fixed and the corresponding Opt_hints_table doesn't exist (or it
+    exists and is fixed) then there's nothing to do, so return early.
+  */
+  if (is_fixed() && (!tab || tab->is_fixed()))
+     return;
+
+  /*
+    This instance will have been marked as fixed on the basis of its
+    attachment to a SELECT_LEX (during get_qb_hints) but that is
+    insufficient to consider it fixed for the case where a TABLE
+    instance is required but not yet available.  If the associated
+    table isn't yet fixed, then fix this hint as though it were unfixed.
+  */
+  DBUG_ASSERT(!table_list->opt_hints_table);
+  DBUG_ASSERT(tab);
+  DBUG_ASSERT(tab->child_array_ptr()->size() == 0);
+  table_list->opt_hints_qb= this;
+  table_list->opt_hints_table= tab;
+  set_fixed();
```

Comment:
Are you sure this is correct? Once again we are processing a single table but `set_fixed()` for the whole query block.